From 364ff4212ecddef27a4c4f28b0cc122ce10fdfcf Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Fri, 28 Jun 2024 10:03:59 +0200
Subject: [PATCH 0961/1695] drivers: stm32_risab: clear SRWIAD bit if property
 not present in DT

Clear SRWIAD bit in RISAB that allows secure read/write operations on
non secure memory protected by RISABs if the property is not present in
the DT. Earlier boot stages may have set this bit.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 1d4591b40fd7ea2df54f965e2024a59bdd79ba39 drivers: stm32_risab: clear SRWIAD bit if property not present in DT]
Change-Id: I605fea5b08f7abf6694c09a6a2b231ad38d74f31
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/390286
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/firewall/stm32_risab.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/core/drivers/firewall/stm32_risab.c b/core/drivers/firewall/stm32_risab.c
index f5508e933..acd815af3 100644
--- a/core/drivers/firewall/stm32_risab.c
+++ b/core/drivers/firewall/stm32_risab.c
@@ -525,10 +525,12 @@ static TEE_Result parse_dt(const void *fdt, int node,
 	return TEE_SUCCESS;
 }
 
-static void set_srswiad_conf(struct stm32_risab_pdata *risab_d)
+static void configure_srwiad(struct stm32_risab_pdata *risab_d)
 {
 	if (risab_d->srwiad)
 		io_setbits32(risab_base(risab_d), _RISAB_CR_SRWIAD);
+	else
+		io_clrbits32(risab_base(risab_d), _RISAB_CR_SRWIAD);
 };
 
 static void clean_iac_regs(struct stm32_risab_pdata *risab_d)
@@ -805,7 +807,7 @@ static TEE_Result stm32_risab_pm_resume(struct stm32_risab_pdata *risab)
 	if (is_tdcid) {
 		if (risab->base.pa == RISAB6_BASE)
 			set_vderam_syscfg(risab);
-		set_srswiad_conf(risab);
+		configure_srwiad(risab);
 		clean_iac_regs(risab);
 	}
 
@@ -921,7 +923,7 @@ static TEE_Result stm32_risab_probe(const void *fdt, int node,
 		if (risab_d->base.pa == RISAB6_BASE)
 			set_vderam_syscfg(risab_d);
 		clean_iac_regs(risab_d);
-		set_srswiad_conf(risab_d);
+		configure_srwiad(risab_d);
 	}
 
 	apply_rif_config(risab_d);
-- 
2.47.2

